src/exceptions.rs

```diff
@@ -9,24 +9,65 @@ use aarch64::{mask, restore};
 
 use core::marker::PhantomData;
 
+/// Scope guard for exception-free sections.
+///
+/// This restores the previous mask state when it is dropped, even if something panics in the
+/// meantime (if panics are unwound).
+///
+/// We don't expose this in the crate API because if scope guards are dropped in the wrong order
+/// then the mask state won't be properly restored.
+#[cfg(target_arch = "aarch64")]
+struct ExceptionGuard {
```

Comment:
Move to within body of `exception_free`?

Reply:
I think nesting it there makes it harder to read, and doesn't gain anything as it's already private to this module.

Reply:
Fair enough